File: server/blender_live.ts
// Talking to a Blender that is already open.
//
// The batch path spawns Blender per call: about a second of process start
// before a single question is answered, and several for a heavy file. That
// cost is what makes batching feel obligatory -- and batching is what forces
// every decision to be made blind and up front.
//
// A live session answers a measurement in ~17ms and a change-and-look in
// ~100ms, measured. At that price, act-then-look is simply how you work, which
// is the whole difference between driving an application and posting it a
// letter.
import fs from "node:fs/promises"
import net from "node:net"
import os from "node:os"
import path from "node:path"
import type { BlenderResult } from "./blender.ts"

/** One socket per project, under the user's own runtime directory. */
export function socketPath(project: string): string {
  const base = process.env.XDG_RUNTIME_DIR || path.join(os.tmpdir(), `omatron-${process.getuid?.()}`)
  return path.join(base, `omatron-blender-${project}.sock`)
}

export async function isLive(project: string): Promise<boolean> {
  const p = socketPath(project)
  try {
    await fs.access(p)
  } catch {
    return false
  }
  // A socket file outlives the process that made it, so its presence proves
  // nothing. Connecting does.
  return await new Promise<boolean>((resolve) => {
    const c = net.createConnection(p)
    const done = (v: boolean) => { try { c.destroy() } catch {} ; resolve(v) }
    c.once("connect", () => done(true))
    c.once("error", () => done(false))
    setTimeout(() => done(false), 700)
  })
}

/** Send one batch to the live session and wait for its answer. */
export function send(
  project: string,
  request: Record<string, unknown>,
  timeoutMs = 300_000,
): Promise<BlenderResult> {
  return new Promise((resolve, reject) => {
    const c = net.createConnection(socketPath(project))
    let buf = ""
    const fail = (e: Error) => { try { c.destroy() } catch {} ; reject(e) }
    const timer = setTimeout(
      () => fail(new Error("the live Blender session did not answer in time")), timeoutMs)

    c.on("connect", () => c.write(JSON.stringify({ ...request, timeout: timeoutMs / 1000 }) + "\n"))
    c.on("data", (d) => {
      buf += d.toString()
      const nl = buf.indexOf("\n")
      if (nl < 0) return
      clearTimeout(timer)
      try {
        resolve(JSON.parse(buf.slice(0, nl)) as BlenderResult)
      } catch (e) {
        fail(new Error(`the live session sent something unreadable: ${String(e)}`))
      }
      try { c.destroy() } catch {}
    })
    c.on("error", (e) => { clearTimeout(timer); fail(e as Error) })
  })
}

/** Stop a live session, leaving the .blend as it stands. */
export async function stop(project: string): Promise<boolean> {
  const p = socketPath(project)
  try {
    await send(project, { ops: [], quit: true }, 5_000)
  } catch {}
  try { await fs.unlink(p) } catch {}
  return true
}

File: server/blender_live.py
# A Blender that stays open and takes instructions -- the session, not the batch.
#
# The batch path (blender_ops.py run headless) opens a file, applies a list,
# saves and exits. That structure forces everything to be decided up front: you
# cannot look part-way and change your mind, so a batch is a bet placed blind,
# and a wrong one costs the whole build.
#
# This is the other shape, and it is the one the browser tools already have.
# Open once, then act and look, act and look, against the thing that is
# actually there. A person modelling does exactly this; the only difference
# here is that the looking can be a measurement as well as a glance.
#
# IT EXECUTES THE SAME OP TABLE. blender_ops is imported, not reimplemented,
# so the interactive path opens no execution surface the batch path did not
# already have. Nothing here evals a string it was handed either.
import bpy
import json
import os
import queue
import socket
import sys
import threading
import traceback

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import blender_ops as ops                                    # noqa: E402

SOCKET_PATH = sys.argv[sys.argv.index("--") + 1]
BLEND_PATH = sys.argv[sys.argv.index("--") + 2]

_work = queue.Queue()


def _execute(req):
    """Run a batch. On the MAIN THREAD -- bpy is not thread-safe, and calling
    it from the socket thread does not raise, it corrupts."""
    ops.MEASURED.clear()
    ops.CHECKPOINTS.clear()
    done, errors = [], []
    for o in req.get("ops", []):
        fn = ops.OPS.get(o.get("op"))
        if fn is None:
            errors.append(f"unknown operation {o.get('op')!r}; "
                          f"known: {', '.join(sorted(ops.OPS))}")
            break
        try:
            done.append(f"{o['op']} -> {fn(o)}")
        except Exception as e:                               # noqa: BLE001
            errors.append(f"{o.get('op')}: {e}")
            break

    # Saved after every successful batch, so the file on disk always matches
    # what is on screen -- and so a crash costs the last operation, not the
    # session. The batch path's read-only rule applies here too.
    readonly = {"measure", "assert", "look", "render_view", "viewport", "view_angle"}
    if not errors and any(o.get("op") not in readonly for o in req.get("ops", [])):
        try:
            bpy.ops.wm.save_as_mainfile(filepath=BLEND_PATH)
        except RuntimeError as e:
            errors.append(f"could not save: {e}")

    return {"applied": done, "errors": errors, "scene": ops.scene(),
            "measured": ops.MEASURED, "checkpoints": ops.CHECKPOINTS,
            "render": None}


def _pump():
    """Drain one request per tick, on the main thread."""
    try:
        req, reply = _work.get_nowait()
    except queue.Empty:
        return 0.05
    try:
        reply.put(_execute(req))
    except Exception:                                        # noqa: BLE001
        reply.put({"applied": [], "errors": [traceback.format_exc(limit=3)],
                   "scene": [], "measured": [], "checkpoints": [], "render": None})
    return 0.01


def _client(conn):
    f = conn.makefile("rwb")
    try:
        line = f.readline()
        if not line:
            return
        req = json.loads(line.decode())
        reply = queue.Queue()
        _work.put((req, reply))
        resp = reply.get(timeout=float(req.get("timeout", 300)))
        f.write((json.dumps(resp) + "\n").encode())
        f.flush()
    except Exception:                                        # noqa: BLE001
        try:
            f.write((json.dumps({"applied": [], "errors": [traceback.format_exc(limit=2)],
                                 "scene": [], "measured": [], "checkpoints": [],
                                 "render": None}) + "\n").encode())
            f.flush()
        except OSError:
            pass
    finally:
        try:
            conn.close()
        except OSError:
            pass


def _serve():
    # A UNIX socket, not a TCP port.
    #
    # It is reachable only through a path in the filesystem, so it carries the
    # owner's permissions -- 0600, this user and no other. A TCP port on
    # localhost is open to every process on the machine and, if anything ever
    # binds it to the wrong interface, to the network as well. That mistake is
    # exactly what the other Blender MCP server shipped.
    try:
        os.unlink(SOCKET_PATH)
    except OSError:
        pass
    srv = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    srv.bind(SOCKET_PATH)
    os.chmod(SOCKET_PATH, 0o600)
    srv.listen(4)
    print(f"OMATRON_LIVE listening on {SOCKET_PATH}")
    while True:
        try:
            conn, _ = srv.accept()
        except OSError:
            break
        threading.Thread(target=_client, args=(conn,), daemon=True).start()


threading.Thread(target=_serve, daemon=True).start()
bpy.app.timers.register(_pump, first_interval=0.1, persistent=True)
